pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -431,6 +443,24 @@ func registerHandlers(informer cache.SharedIndexInformer, q queue.Instance, otyp
 		})
 }
 
+// tryGetLatestObject attempts to fetch the latest version of the object from the cache.
+// Changes may have occurred between queuing and processing.
+func tryGetLatestObject(informer cache.SharedIndexInformer, obj interface{}) interface{} {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		log.Warnf("failed creating key for informer object: %v", err)
+		return obj
+	}
+
+	latest, exists, err := informer.GetIndexer().GetByKey(key)
+	if !exists || err != nil {
+		log.Warnf("couldn't find %q in informer index", key)
+		return obj
```

Comment:
If we have the following sequence:

v1,v2,delete

We will end up processing
v2,v2,delete,v2,v2,delete

So there is some flip flop. I think this is fine, but it brings up the question do we even need `tryGetLatestObject` if we haven't fully solved the "flip flopping"?

Reply:
This does not matter much, this is the eventual consistent model.

Reply:
I think it's still an improvement. If we have the events in the order: v1, v2
We'll process v2 (during SyncAll), then v2, v2 when we start processing the queue.

I think in the case you described, the second set of `v2, v2` would result in `!exists` or the object in the index would have something indicated that the latest state is `deleted`. The events are based on changes to the indexer so the value there should always win. 

If we think we see multiple successive changes to the same resource in a short enough time span, I'd be more inclined to implement workqueue and dedup.

Reply:
Ah actually I totally forgot that getting deletes or multiple updates during this time is extremely rare - only can happen if they occur right when Istiod is starting up. So because of that we need to ensure we are correct during these cases, but optimizing for it is probably premature. so LGTM